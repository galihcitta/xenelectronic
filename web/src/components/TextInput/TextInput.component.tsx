import React, { ChangeEvent, useState } from 'react';

import {
  Wrapper,
  Input,
  Label,
  Error,
  ValidationIcon
} from './TextInput.styles';

export interface TextInputProps {
  name: string;
  id: string;
  placeholder: string;
  value: string;
  label: string;
  type?: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;

  // validation props
  errorMsg: string;
  touched: boolean;
  invalid: boolean;
  shouldValidate: {} | null;
}

export const TextInput: React.FC<TextInputProps> = ({
  name,
  id,
  placeholder,
  value,
  label,
  type = 'text',
  onChange,
  errorMsg,
  touched,
  invalid,
  shouldValidate
}) => {
  const [showError, setShowError] = useState<boolean>(false);

  let isInvalid = null;
  if (invalid && shouldValidate && touched) isInvalid = true;

  const checkIfError = () => {
    if (errorMsg) setShowError(true);
  };

  return (
    <Wrapper>
      <Input
        type={type}
        id={id}
        placeholder={placeholder}
        name={name}
        value={value}
        onChange={onChange}
        onBlur={checkIfError}
      />
      <ValidationIcon
        data-testid="ValidationIcon"
        isInvalid={isInvalid}
        touched={touched}
      >
        {isInvalid ? (
          <i className="fas fa-exclamation-circle"></i>
        ) : (
          <i className="fas fa-check-circle"></i>
        )}
      </ValidationIcon>
      <Label shrink={value.length ? true : false} htmlFor={id}>
        {label}
      </Label>
      <Error showError={showError && invalid && errorMsg !== ''}>
        {errorMsg}
      </Error>
    </Wrapper>
  );
};
