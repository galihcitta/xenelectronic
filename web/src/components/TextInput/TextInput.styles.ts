import styled from 'styled-components';

import { color, fontWeight } from '../../utils/styles';

interface ILabelProps {
  shrink: boolean;
}

interface IIconProps {
  isInvalid: boolean | null;
  touched: boolean;
}

interface IErrorProps {
  showError: boolean;
}

export const Wrapper = styled.p`
  position: relative;
  width: 100%;
`;

export const Label = styled.label`
  color: ${color.gray};
  font-weight: ${fontWeight.regular};
  position: absolute;
  pointer-events: none;
  left: 5px;
  top: ${(props: ILabelProps) => (props.shrink ? '16px' : '42px')};
  font-size: ${(props: ILabelProps) => (props.shrink ? '12px' : '.9rem')};
  transition: 300ms ease all;
`;

export const Input = styled.input`
  width: 100%;
  height: 40px;
  padding: 10px 10px 10px 5px;
  padding-right: 50px;
  border: none;
  border-bottom: 1px solid ${color.gray};
  font-size: 1rem;
  background: none;
  color: ${color.gray};
  margin-top: 30px;
  margin-bottom: 7px;

  &:focus {
    outline: none;
  }

  &:focus ~ ${Label} {
    top: 16px;
    font-size: 12px;
  }

  &[type='password'] {
    letter-spacing: 0.3em;
  }
`;

export const ValidationIcon = styled.span`
  display: block;
  opacity: ${(props: IIconProps) => (props.touched ? '1' : '0')};
  transition: all 0.25s ease-in-out;
  position: absolute;
  top: 39px;
  right: 0;
  font-size: 1.4rem;
  color: ${(props: IIconProps) =>
    props.isInvalid === true ? '#ff4a4a' : '#4ec76e'};
`;

export const Error = styled.span`
  display: block;
  font-size: 0.8rem;
  font-weight: ${fontWeight.semiBold};
  color: ${color.white};
  line-height: 40px;
  width: 100%;
  background: #ff4a4a;
  padding: 0 0.9em;
  border-radius: 11px;
  height: ${(props: IErrorProps) => (props.showError ? '40px' : '0')};
  opacity: ${(props: IErrorProps) => (props.showError ? '1' : '0')};
  transition: all 0.15s ease-in-out;
`;
