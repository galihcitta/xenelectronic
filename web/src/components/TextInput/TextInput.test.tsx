import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { TextInput, TextInputProps } from './TextInput.component';

describe('<TextInput />', () => {
  const mockOnChange = jest.fn();

  const props: TextInputProps = {
    name: 'email',
    id: 'email',
    placeholder: '',
    value: '',
    label: 'Enter email',
    type: 'email',
    onChange: mockOnChange,
    errorMsg: '',
    touched: false,
    invalid: true,
    shouldValidate: {
      required: true,
      isEmail: true
    }
  };

  const setup = (props: TextInputProps) => {
    const utils = render(<TextInput {...props} />);
    const validationIcon = utils.getByTestId('ValidationIcon');
    const input = utils.container.querySelector(
      'input[type="email"]'
    ) as HTMLInputElement;

    return { ...utils, validationIcon, input };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `onChange` func', () => {
    const { input } = setup(props);

    fireEvent.change(input, { target: { value: 'johndoe@gmail.com' } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  it('contains `color:#4ec76e` style rule in `ValidationIcon` element', () => {
    const { validationIcon } = setup(props);
    expect(validationIcon).toHaveStyle('color: #4ec76e');
  });

  describe('when `touched` prop is equal to true', () => {
    beforeEach(() => {
      props.touched = true;
    });

    it('contains `color:#ff4a4a` style rule in `ValidationIcon` element', () => {
      const { validationIcon } = setup(props);
      expect(validationIcon).toHaveStyle('color: #ff4a4a');
    });

    it('contains `opacity: 1` style rule in `ValidationIcon` element', () => {
      const { validationIcon } = setup(props);
      expect(validationIcon).toHaveStyle('opacity: 1');
    });

    it('contains correct icon in `ValidationIcon` element', () => {
      const { validationIcon } = setup(props);
      expect(validationIcon.firstChild).toHaveClass('fa-exclamation-circle');
    });
  });

  it('contains correct style rules in `Label` element when `value` prop is not empty', () => {
    const { container } = setup({ ...props, value: 'johndoe@gmail.com' });
    const label = container.querySelector('label');

    expect(label).toHaveStyle('top: 16px');
    expect(label).toHaveStyle('font-size: 12px');
  });

  describe('when `errorMsg` prop is not empty', () => {
    beforeEach(() => {
      props.errorMsg = 'Invalid email';
    });

    it('contains correct style rules in `Error` element', () => {
      const { input, getByText } = setup(props);
      fireEvent.blur(input);

      expect(getByText(/invalid email/i)).toHaveStyle('height: 40px');
      expect(getByText(/invalid email/i)).toHaveStyle('opacity: 1');
    });

    it('contains correct text in `Error` element', () => {
      const { getByText } = setup(props);
      expect(getByText(/invalid email/i)).toBeInTheDocument();
    });
  });
});
