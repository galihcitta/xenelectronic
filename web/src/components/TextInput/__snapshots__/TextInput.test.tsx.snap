// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> renders correctly 1`] = `
<DocumentFragment>
  <p
    class="sc-bdVaJa jGgciL"
  >
    <input
      class="sc-htpNat kvebgI"
      id="email"
      name="email"
      placeholder=""
      type="email"
      value=""
    />
    <span
      class="sc-bxivhb BciNW"
      data-testid="ValidationIcon"
    >
      <i
        class="fas fa-check-circle"
      />
    </span>
    <label
      class="sc-bwzfXH fvEVNp"
      for="email"
    >
      Enter email
    </label>
    <span
      class="sc-ifAKCX iHjGSE"
    />
  </p>
</DocumentFragment>
`;
